odl/contrib/tensorflow.py

```diff
@@ -0,0 +1,512 @@
+                        differentiable=True):
+    """Convert `Operator` or `Functional` into tensorflow layer.
+
+    Parameters
+    ----------
+    odl_op : `Operator` or `Functional`
+        The operator that should be wrapped to a tensorflow layer.
+    default_name : str
+        Default name for tensorflow layers created.
+    differentiable : boolean
+        True if the layer should be differentiable, in which case  ``odl_op``
+        should implement `Operator.derivative` which in turn implements
+        `Operator.adjoint`, it is properly wrapped in ``tensorflow_layer``, and
+        gradients propagate as expected.
+
+        Otherwise assumes that the derivative is everywhere zero.
+
+    Returns
+    -------
+    tensorflow_layer : callable
+        Callable that, when called with an `tensorflow.Tensor` of shape
+        `(n, *odl_op.domain.shape, 1)` where ``n`` is the number of batches
+        returns another `tensorflow.Tensor`.
+
+        If ``odl_op`` is an `Operator`, the shape of the returned tensor is
+        `(n, *odl_op.range.shape, 1)`.
+
+        Hence for each evaluation, ``odl_op`` is called a total of ``n`` times.
+
+        The `dtype` of the tensor is the same as the respective ODL spaces.
+
+    """
+    if not TENSORFLOW_AVAILABLE:
+        raise ImportError('Tensorflow not installed')
+
+    from tensorflow.python.framework import ops
+
+    def py_func(func, inp, Tout, stateful=True, name=None, grad=None):
+        """Define custom py_func which takes also a grad op as argument.
+
+        We need to overwrite this function since the default tensorflow py_func
+        does not support custom gradients.
+
+        Parameters
+        ----------
+        func : callable
+            Python function that takes and returns numpy arrays.
+        inp : sequence of `tensorflow.Tensor`
+            Input tensors for the function
+        Tout : sequence of `tensorflow.dtype`
+            Datatype of the output(s).
+        stateful : bool
+            If the function has internal state, i.e. if calling the function
+            with a given input always gives the same output.
+        name : string
+            Name of the python function
+        grad : callbable
+            Gradient of the function.
+        """
+        if grad is None:
+            return tf.py_func(func, inp, Tout, stateful=stateful, name=name)
+        else:
+            if stateful:
+                override_name = 'PyFunc'
+            else:
+                override_name = 'PyFuncStateless'
+
+            # Need to generate a unique name to avoid duplicates:
+            rnd_name = override_name + 'Grad' + str(uuid.uuid4())
+
+            tf.RegisterGradient(rnd_name)(grad)
+            g = tf.get_default_graph()
+
+            with g.gradient_override_map({override_name: rnd_name}):
+                return tf.py_func(func, inp, Tout, stateful=stateful,
+                                  name=name)
```

Comment:
Yes it is unknown and on the issue tracker but apparently was closed:

https://github.com/tensorflow/tensorflow/issues/1095